### PR TITLE
feat(prisma): @unthrown/prisma — a Prisma Client extension bridging queries into AsyncResult

### PR DESCRIPTION
**Pull request by btravers**
## What

A new integration package, **`@unthrown/prisma`** (`packages/prisma`), following the `@unthrown/<peer>` convention: a shareable Prisma Client extension (`Prisma.defineExtension` from `@prisma/client/extension`) that adds `try`-prefixed variants of the model delegate operations **alongside** the raw promise ones.

```ts
const db = new PrismaClient({ adapter }).$extends(unthrownPrisma);

const users = db.user.tryFindMany({ select: { id: true } });
//    ^? AsyncResult<{ id: number }[], DriverError>
```

## Design

- **Per-operation error unions** — the key feature over one flat union: reads (`tryFindMany` / `tryFindUnique` / `tryCount`) can only fail with `DriverError`; writes add `UniqueConstraintViolation` (P2002) and `ForeignKeyViolation` (P2003); `tryFindUniqueOrThrow` / `tryUpdate` / `tryDelete` add `RecordNotFound` (P2025). Everything else folds into `DriverError` with the cause preserved.
- **Payload inference survives the wrap** — typed with the official `Prisma.Args` / `Prisma.Result` / `Prisma.Exact` utilities (the `$allModels` pattern); a naive mapped type over the client would destroy `select`/`include` narrowing.
- **Qualification once, at the boundary (Thesis #3)** — `fromPromise(…, qualifyPrismaError)` inside the extension; no raw Promise reaches the caller. P2002's offending `fields` are extracted from both the classic `meta.target` and the driver-adapter shape (`meta.driverAdapterError.cause.constraint.fields`).
- **`$tryTransaction`** — an interactive transaction whose callback speaks `AsyncResult`: an `Err` is thrown internally as a sentinel so Prisma rolls back, then re-surfaced as the same typed `Err`; a defect also rolls back and **stays a defect** (re-minted through `fromPromise`'s `defect` factory). The `try*` methods propagate onto `tx`; the deny list removes `$tryTransaction` itself, so nesting is a compile error.
- **The raw methods stay on purpose** — they are the escape hatch for batch `$transaction([...])`, which needs unexecuted `PrismaPromise`s.

## Verification

- **23 runtime tests against a real client** over in-memory SQLite (`@prisma/adapter-better-sqlite3`): every mapped P-code is provoked for real (duplicate email → P2002, dangling relation → P2003, missing row → P2025), and `$tryTransaction` covers commit, rollback-on-`Err`, rollback-on-defect, and a transaction-level failure that bypasses the sentinel. **100% statement / branch / function / line coverage**, enforced by thresholds.
- **Type-level tests** (`types.test-d.ts`, checked by the package's regular `tsc --noEmit`): exact `select` narrowing, per-operation error unions, `Prisma.Exact` arg rejection, transaction error-union and deny list.
- Dogfoods `@unthrown/vitest` (`toBeOkWith` / `toBeErrTagged` / `toBeDefect`).

## Housekeeping

- Catalog entries for `prisma` / `@prisma/client` / `@prisma/adapter-better-sqlite3` / `better-sqlite3` (all past the 7-day supply-chain gate); `allowBuilds` for better-sqlite3's native bindings, engines disabled (Prisma 7 is WASM); `react`/`react-dom` peer-ignore for the bundled Prisma Studio we never run.
- `@unthrown/prisma` versions **independently** (left the fixed group in a review round — its majors track `@prisma/client`'s cadence, not the family's); changeset included (minor, starts at 0.1.0).
- Registered in the docs pipeline (copy-docs, VitePress sidebar) and the root README package table.
- oxlint/knip exclude the generated test client (`src/generated`, gitignored, regenerated by the test/typecheck scripts — the published package is schema-agnostic and ships `dist` only).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Update: cursor pagination (`tryPaginate`)

Added \`tryPaginate(query).withCursor({ limit, after, before, getCursor, parseCursor })\` → \`AsyncResult<[results, meta], DriverError>\` — the [prisma-extension-pagination](https://github.com/deptyped/prisma-extension-pagination) cursor API (same option names, same \`[results, meta]\` shape), with three changes:

- **Folds in deptyped/prisma-extension-pagination#36** (fixing [#35](https://github.com/deptyped/prisma-extension-pagination/issues/35), never merged upstream): a cursor pointing at a record that no longer matches the query filter must not skip the first element of the page. Instead of \`skip: 1\`, the page is over-fetched by two and the boundary row is dropped only when it round-trips to the request cursor — now covered by the tests the upstream PR was missing, in both directions (\`after\` / \`before\`) and at both over-fetch depths.
- **\`before\` + \`limit: null\` is a compile error** — Prisma's negative \`take\` cannot express "everything before the cursor"; upstream silently pages the wrong way for this combination.
- **The default \`parseCursor\` preserves the id's type** — all-digits cursors parse back to numbers (autoincrement ids), anything else stays a string (uuid/cuid), where upstream hard-assumes numeric ids.

The query args exclude \`cursor\` / \`take\` / \`skip\` at the type level (pagination owns them), \`select\`/\`include\` payload inference flows through, and a throwing \`parseCursor\` (malformed cursor from an API client) surfaces as a modeled \`DriverError\`, never a raw throw. 37 tests, still 100% statement/branch/function/line coverage.

### CODE REVIEW

File: packages/prisma/src/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+  const adapter = new PrismaBetterSqlite3({ url: ":memory:" });
+  const db = new PrismaClient({ adapter }).$extends(unthrownPrisma);
+  open.push(db);
+  for (const stmt of DDL) await db.$executeRawUnsafe(stmt);
+  return db;
+};
+
+afterEach(async () => {
+  await Promise.all(open.splice(0).map((db) => db.$disconnect()));
+});
```
Comment (btravers) [inline packages/prisma/src/index.spec.ts]: use vitest context to create a fixture

Comment (btravers) [reply packages/prisma/src/index.spec.ts]: Done in 2d0547f — `test.extend` fixtures: `db` provisions the client + DDL and disconnects on teardown, `seededDb` layers the six pagination users on top of it. Lazy, so each test only pays for what it destructures; the `makeDb` helper, the open-clients array, and the `afterEach` are gone. (One inline `no-empty-pattern` disable — Vitest syntactically requires the destructuring pattern on a dependency-less fixture.)

File: packages/prisma/src/pagination.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+// page. Instead of `skip: 1` past the cursor row — which skips a real result
+// when the cursor row itself is filtered out — the page is over-fetched by two
+// and the boundary row is dropped only when it IS the cursor row.
+
+export type CursorPaginationMeta = {
+  hasPreviousPage: boolean;
+  hasNextPage: boolean;
+  startCursor: string | null;
+  endCursor: string | null;
+};
```
Comment (btravers) [inline packages/prisma/src/pagination.ts]: use union type

in deed, endCusor is string if hasNextPage is true and null if hasNextPage is false

same reasoning for startCursor with hasPreviousPage

File: .changeset/config.json [hunk 3]
```diff
@@ -11,7 +11,8 @@
       "@unthrown/neverthrow",
       "@unthrown/boxed",
       "@unthrown/oxlint",
-      "@unthrown/standard-schema"
+      "@unthrown/standard-schema",
+      "@unthrown/prisma"
```
Comment (btravers) [inline .changeset/config.json]: as prisma can evolve quickly, is it a good idea to follow the same version as other libs from the workspace?

File: packages/prisma/prisma/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// (gitignored) by the test/typecheck scripts.
+
+generator client {
+  provider = "prisma-client"
+  output   = "../src/generated/prisma"
+}
+
+datasource db {
+  provider = "sqlite"
+}
```
Comment (btravers) [inline packages/prisma/prisma/schema.prisma:13]: as postgres is the most popular database, should we consider testing the package with postgres using testcontainers?

File: packages/prisma/src/pagination.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  options: RuntimeOptions,
+): Promise<[unknown[], CursorPaginationMeta]> => {
+  const { limit, after, before } = options;
+  const getCursor = (options.getCursor ?? defaultGetCursor) as (row: unknown) => string;
+  const parseCursor = options.parseCursor ?? defaultParseCursor;
+
+  // Is this row the cursor row? Compared through the cursor round-trip, so it
+  // works for any serialization the caller chose.
+  const sameCursor = (row: unknown, cursor: unknown): boolean =>
+    JSON.stringify(parseCursor(getCursor(row))) === JSON.stringify(cursor);
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/prisma/src/pagination.ts]: Fixed in 4a233ca — `sameCursor` now uses a structural deep-equal: key-order independent, and safe for `bigint`, `Date`, nested composite-key records (and `Uint8Array`, via the record path). Covered by three new engine-level tests: bigint ids round-trip (stringify would throw), a parseCursor emitting keys in a different order per call (stringify would false-negative and leave the cursor row in the page), and Date + array values comparing by value.

File: packages/prisma/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+//   const db = new PrismaClient({ adapter }).$extends(unthrownPrisma);
+//
+//   await db.user.tryCreate({ data }).match({ ok, err, defect });
+//   // err is UniqueConstraintViolation | ForeignKeyViolation | DriverError
+//
+// The raw promise methods stay available on purpose: they are the escape hatch
+// for batch `$transaction([...])`, which needs unexecuted `PrismaPromise`s.
+
+import { Prisma } from "@prisma/client/extension";
+import { PrismaClientKnownRequestError } from "@prisma/client/runtime/client";
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/prisma/src/index.ts]: Checked against the installed 7.8.0: `@prisma/client/runtime/library` does **not** exist in Prisma 7 — it was the v4–v6 library-engine entry; `runtime/client` is the v7 runtime and exactly what the generated client imports. But the underlying concern is real, so 4a233ca removes the runtime import from the package entirely: `qualifyPrismaError` now recognizes the error structurally (`name === "PrismaClientKnownRequestError"` + string `code`), which is stable across every runtime path AND immune to dual-copy `instanceof` failures.

File: packages/prisma/src/index.spec.ts [hunk 7]
```diff
@@ -0,0 +1,402 @@
+// Integration tests against a REAL Prisma client over in-memory SQLite
+// (`@prisma/adapter-better-sqlite3`): every mapped P-code is provoked for real —
+// a duplicate email for P2002, a dangling relation for P2003, a missing row for
+// P2025 — and `$tryTransaction` is exercised end-to-end (commit, rollback on
+// `Err`, rollback on defect, `try*` methods available on the itx client). The
+// pure P-code mapping gets its own unit block.
+
+import { PrismaBetterSqlite3 } from "@prisma/adapter-better-sqlite3";
+import { PrismaClientKnownRequestError } from "@prisma/client/runtime/client";
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/prisma/src/index.spec.ts:9]: Kept in the spec deliberately: the test constructs REAL error instances, and `runtime/client` is the same module the generated test client uses (`runtime/library` doesn't exist in v7). The spec exercising the real class also guards the production code's structural check against drift — if Prisma renamed the class, these tests would fail.

File: packages/prisma/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+     * ```
+     */
+    $tryTransaction<C, T, E>(
+      this: C,
+      fn: (tx: Omit<C, TxDenyList>) => AsyncResult<T, E>,
+      options?: {
+        maxWait?: number;
+        timeout?: number;
+        isolationLevel?: string;
+      },
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/prisma/src/index.ts:537]: Partially done in 4a233ca. The schema-derived `Prisma.TransactionIsolationLevel` is generated per-database (sqlite has only `Serializable`) and is not nameable from a shareable extension — even the `@prisma/client/extension` entry types `$transaction`'s `isolationLevel` as plain `string`. So the extension now exports a closed union of all levels Prisma accepts (`ReadUncommitted | ReadCommitted | RepeatableRead | Snapshot | Serializable`): typos are compile errors; a level your database doesn't support still fails at runtime as a `DriverError`. Type-level test included.

File: packages/prisma/src/pagination.ts (hunk 2 above)
Comment (btravers) [reply packages/prisma/src/pagination.ts]: Union type done in 4a233ca — but coupled to **page emptiness**, not the flags, because the flag coupling doesn't hold: the last page has `hasNextPage: false` with a non-null `endCursor` (its rows exist — see the "reaches the last page" test: `endCursor: "6"`), and an empty page past the end has `hasPreviousPage: true` with a null `startCursor`. The real invariant is that the boundary cursors travel together — `{ startCursor: string; endCursor: string } | { startCursor: null; endCursor: null }` — so checking one narrows the other. Type-level test included.

File: .changeset/config.json (hunk 3 above)
Comment (btravers) [reply .changeset/config.json]: Agreed — done in 4a233ca: `@unthrown/prisma` is out of the fixed group and versions independently, starting at 0.1.0 via the existing changeset. Its majors should track `@prisma/client`'s cadence (peer bumps `^7` → `^8`), not drag the whole family; and conversely an unthrown-family major shouldn't force a meaningless major here.

File: packages/prisma/prisma/schema.prisma (hunk 4 above)
Comment (btravers) [reply packages/prisma/prisma/schema.prisma:13]: Worth doing, as a follow-up PR rather than here. The P2002 `meta` shape already differs by adapter (the driver-adapter nesting this PR handles), so Postgres coverage has real value — pg is where `meta.target`, FK codes, and the richer isolation levels live. Suggested shape: a separate `postgres.spec.ts` excluded from the default vitest include + a `test:postgres` script using `@testcontainers/postgresql`, run as an extra CI job — keeping the default `pnpm test` hermetic (no Docker required locally). Happy to open it once this lands.

**Comment by btravers** [comment]
## Review — `@unthrown/prisma`

Rebased onto `main` (clean; then resolved to keep the branch current) and reviewed the whole surface. **Strong package — I'd merge it after you weigh the two notes below.** Verified locally: `typecheck`, **40 tests** (real in-memory SQLite), `build`, `lint`, and `build:docs` all green.

### What stands out
- **Thesis-clean boundary** — `fromPromise(…, qualifyPrismaError)` lives *inside* the extension, so no raw `Promise` (and so no un-triaged rejection) ever reaches a combinator (Thesis #3). ✔
- **Per-operation error unions** are the real feature: a read is typed `DriverError`-only, writes add exactly the constraint violations their SQL can raise. Sound — the runtime qualify produces the full union, but each op can only actually hit its declared subset, so the down-casts are honest.
- **Version-robust P-code recognition** — structural (`name` + string `code`) instead of `instanceof` against a Prisma runtime class whose module path moves across v4–v7. P2002 `fields` are read from *both* `meta.target` and the driver-adapter shape. Nice.
- **`$tryTransaction`** — `Err` → sentinel throw → rollback → re-surfaced as the same typed `Err`; a **defect rolls back and stays a defect** (re-minted through the injected `defect` factory); nesting is a compile error via the deny list. Correct and careful.
- **Pagination** folds in the never-merged upstream fix ([#35](https://github.com/deptyped/prisma-extension-pagination/issues/35)) with real tests in both directions; structural cursor equality survives `bigint`/`Date`/array/key-order; `before` + `limit: null` is a compile error; the default `parseCursor` preserves the id's type.
- **Tests**: every mapped P-code provoked for real, transaction commit/rollback-on-Err/rollback-on-defect/tx-level-failure, and the pagination edge cases — at 100% coverage thresholds. Dogfoods `@unthrown/vitest`.

### Notes (non-blocking)
1. **API surface gap.** No `try*` for `findFirst` / `findFirstOrThrow` / `upsert` / `createMany` / `updateMany` / `deleteMany` / `aggregate` / `groupBy`. `findFirst` especially is everyday. Not a defect — the raw methods remain and `qualifyPrismaError` is exported — but a natural follow-up. (The new guide documents the `fromPromise(raw, qualifyPrismaError)` escape hatch in the meantime.)
2. **`$tryTransaction` sync-throw edge.** If the callback throws *synchronously* before returning its `AsyncResult` (which the type says it shouldn't), that throw is qualified as a `DriverError` — a modeled `Err` — rather than a `Defect`. Arguably a bug in the callback should surface as a defect. Minor; flagging only.

### Changes I pushed on this branch
- **Fixed a gate violation:** `build:docs` was emitting 4 TypeDoc warnings (`CreateError`/`UpdateError`/`DeleteError`/`TxDenyList` referenced-but-not-exported). Silenced via `intentionallyNotExported` (matching core). Alternatively you may want to **export** `CreateError`/`UpdateError`/`DeleteError` as public unions — your call.
- **Added `docs/guide/prisma.md`** (per-operation errors, `matchTags`, `$tryTransaction`, pagination, escape hatch) + wired it into the Ecosystem sidebar; repointed the package README's Documentation link at `/guide/prisma` (was `/guide/interop`, which didn't mention Prisma).

Heads-up on merge order: no overlap with the operator-rename PR (#86) — this package uses none of the renamed operators.

**Comment by btravers** [comment]
**Follow-up on review note 1 (API surface gap) — resolved in e358350.** The extension now wraps **all seventeen** model delegate operations. The ten added:

| new | error channel |
| --- | --- |
| `tryFindFirst` | `DriverError` |
| `tryFindFirstOrThrow` | `RecordNotFound \| DriverError` |
| `tryAggregate` / `tryGroupBy` | `DriverError` |
| `tryCreateMany` / `tryCreateManyAndReturn` | `UniqueConstraintViolation \| ForeignKeyViolation \| DriverError` |
| `tryUpsert` | `UniqueConstraintViolation \| ForeignKeyViolation \| DriverError` (no P2025 — a miss *creates*) |
| `tryUpdateMany` / `tryUpdateManyAndReturn` | `UniqueConstraintViolation \| ForeignKeyViolation \| DriverError` (no P2025 — zero matches is `Ok({ count: 0 })`) |
| `tryDeleteMany` | `ForeignKeyViolation \| DriverError` |

- Each operation exercised against the real SQLite client with its P-codes provoked for real (P2002 in a batch, P2002 through an upsert's update arm, P2003 on `deleteMany` of a referenced parent, `Ok(0)`-not-P2025 for zero-match batches) — **46 tests, coverage still 100/100/100/100**.
- Type-level assertions extended: `select` narrowing through `findFirst`/`createManyAndReturn`/`updateManyAndReturn`, and the exact per-operation unions (incl. the P2025 *exclusions*).
- Guide table, README, changeset, and CLAUDE.md updated; the escape-hatch section now covers only what genuinely remains raw: batch `$transaction([...])` and `$queryRaw`/`$executeRaw`. TypeDoc stays warning-free.

Review note 2 ($tryTransaction sync-throw → `DriverError` instead of defect) is still open — flagging again so it isn't lost; happy to fix it here or as a follow-up.

**Comment by btravers** [comment]
**Review note 2 ( sync-throw) — resolved in 2aa54fa.** A callback that throws instead of returning an `AsyncResult` now rolls back and **stays a defect**, instead of being downgraded to a modeled `DriverError`. The catch is scoped precisely to the callback's own invocation, so rejections from Prisma's BEGIN/COMMIT/timeout machinery still qualify as query failures (the timeout test still passes). New discriminating test included — 47 tests, coverage still 100/100/100/100. Both review notes are now closed.

File: packages/prisma/README.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+// Custom serialization (composite keys, non-id cursors):
+.withCursor({
+  limit: 20,
+  getCursor: ({ postId, userId }) => `${postId}:${userId}`,
+  parseCursor: (cursor) => {
+    const [postId, userId] = cursor.split(":");
+    return { userId_postId: { postId: Number(postId), userId: Number(userId) } };
+  },
+});
```
Comment (btravers) [inline packages/prisma/README.md:86]: not nicely indented

File: packages/prisma/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+// so a raw Promise never escapes. Each public method casts the result down to
+// its per-operation payload and error union (the `$allModels` implementation
+// side is untyped by design; the declared signatures carry the safety).
+type UntypedDelegate = Record<string, (args?: unknown) => Promise<unknown>>;
+
+const query = (self: unknown, op: string, args?: unknown): AsyncResult<unknown, PrismaQueryError> =>
+  fromPromise(
+    (Prisma.getExtensionContext(self) as unknown as UntypedDelegate)[op]!(args),
+    qualifyPrismaError,
+  );
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/prisma/src/index.ts:164]: Fixed in a26a871 — `query()` now hands `fromPromise` a **thunk**, so a synchronously-throwing delegate is absorbed by the boundary (qualified via `qualifyPrismaError` → `DriverError`, per the documented "validation errors land here" bucket) instead of escaping as a raw throw.

File: packages/prisma/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+            ? new Rollback(result.error, false)
+            : new Rollback(result.cause, true);
+        }, options),
+        (cause, defect) =>
+          cause instanceof Rollback
+            ? cause.wasDefect
+              ? defect(cause.carried)
+              : (cause.carried as E | PrismaQueryError)
+            : qualifyPrismaError(cause),
+      );
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/prisma/src/index.ts:569]: Fixed in a26a871 — same treatment: `$tryTransaction` passes a thunk, so a synchronous throw out of `$transaction` itself stays inside the boundary.

File: packages/prisma/src/pagination.ts [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    );
+  }
+  return String(id);
+};
+
+// Preserve the id's type through the round-trip: an all-digits cursor parses
+// back to a number (autoincrement ids), anything else stays a string.
+const defaultParseCursor = (cursor: string): unknown => ({
+  id: /^\d+$/.test(cursor) ? Number(cursor) : cursor,
+});
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/prisma/src/pagination.ts]: Fixed in a26a871 — the default parser now promotes an all-digit cursor beyond `Number.MAX_SAFE_INTEGER` to a `bigint` (`Number.isSafeInteger` gate), so a BigInt id round-trips exactly. Unit-tested at 2^53 + 1, asserting the *exact* bigint reaches Prisma's `cursor` arg. Safe-range ids still parse to plain numbers (Prisma coerces number→BigInt fine).

File: docs/guide/prisma.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+- **A cursor pointing at a now-filtered-out row doesn't skip the first element**
+  of the page (folds in the fix for
+  [deptyped/prisma-extension-pagination#35](https://github.com/deptyped/prisma-extension-pagination/issues/35)).
+- **`before` + `limit: null` is a compile error** — Prisma's negative `take`
+  can't express "everything before the cursor, unbounded."
+- **The default cursor preserves the id's type** — all-digit cursors parse back
+  to numbers (autoincrement ids), anything else stays a string (uuid / cuid).
+  Provide `getCursor` / `parseCursor` for composite keys or a selection without
+  `id`.
```
[the remark this answers is not in the document]
Comment (btravers) [reply docs/guide/prisma.md]: Fixed in a26a871 — with the parser now promoting to `bigint` past `Number.MAX_SAFE_INTEGER`, the guide claim is accurate; reworded to spell out the number/bigint/string behavior and it still points composite keys / id-less selections at `parseCursor`.

File: packages/prisma/README.md (hunk 9 above)
Comment (btravers) [reply packages/prisma/README.md:86]: Fixed in a26a871 — the snippet was a floating `.withCursor(` expression fragment; it's now a complete statement (`const liked = await db.like.tryPaginate(…).withCursor({ … })`), so it indents like real code.
